elyra/util/cos.py

```diff
@@ -28,36 +28,84 @@ class CosClient(LoggingConfigurable):
         super().__init__(**kwargs)
-        if config:
+
+        cred_provider = None
+        if config is None:
+            # The client was invoked by an entity that does not utilize
+            # runtime configurations. Use the explicitly provided parameter
+            # values instead.
+            cred_provider = providers.StaticProvider(
+                access_key=access_key,
+                secret_key=secret_key,
+            )
+            self.endpoint = endpoint
+            self.bucket = bucket
+        else:
+            auth_type = config.metadata['cos_auth_type']
             self.endpoint = urlparse(config.metadata['cos_endpoint'])
-            self.access_key = config.metadata.get('cos_username')
-            self.secret_key = config.metadata.get('cos_password')
             self.bucket = config.metadata['cos_bucket']
-        else:
-            self.endpoint = urlparse(endpoint)
-            self.access_key = access_key
-            self.secret_key = secret_key
-            self.bucket = bucket
+            if auth_type in ['USER_CREDENTIALS', 'KUBERNETES_SECRET']:
+                # TODO remove temporary validation checks after
+                # metadata can properly enforce constraints
+                if len(config.metadata.get('cos_username', '').strip()) == 0 or\
+                   len(config.metadata.get('cos_password', '').strip()) == 0:
+                    raise RuntimeError('Cannot connect to object storage. '
+                                       f'Authentication provider \'{auth_type}\' requires '
+                                       'a username and password. Update runtime configuration '
+                                       f'\'{config.display_name}\' and try again.')
+                if auth_type == 'USER_CREDENTIALS' and\
+                   len(config.metadata.get('cos_secret', '').strip()) > 0:
+                    raise RuntimeError('Cannot connect to object storage. '
+                                       f'Authentication provider \'{auth_type}\' does not '
+                                       'support Kubernetes secrets. Update runtime configuration '
+                                       f'\'{config.display_name}\' and try again.')
+                if auth_type == 'KUBERNETES_SECRET' and\
+                   len(config.metadata.get('cos_secret', '').strip()) == 0:
+                    raise RuntimeError('Cannot connect to object storage. '
+                                       f'Authentication provider \'{auth_type}\' requires '
+                                       'a Kubernetes secret name. Update runtime configuration '
+                                       f'\'{config.display_name}\' and try again.')
+                cred_provider = providers.StaticProvider(
+                    access_key=config.metadata['cos_username'],
+                    secret_key=config.metadata['cos_password'],
+                )
+            elif auth_type == 'AWS_ENV_VARIABLES':
```

Comment:
Is this a valid `cos_auth_type`?  It's not listed in the description.  It too could be embedded in the save hook.

Reply:
It's not, but this is a side effect of the fact that the PR was and not ready for review and still a work in progress.